### ISSUE TITLE
Implement a class to clone checks for MT mode

### ISSUE THREAD

**Issue by soon** [checkstyle/checkstyle#4869]
A check cloning service is required to use checks in different threads. In this issue the following classes and interfaces should be created:

1. CheckCloneService - the main class for creating check clones
1. OneCheckInstancePerThread - the marker interface for thread local checks
1. OneCheckInstancePerApplication - the marker interface for global-stateful and stateless checks

**Comment by romani** [comment]
Sorry for being late on the party

Can somebody explain me "Why we need cloning of Checks ?".
We have config file, we have ability to create Tree of modules as much as we need from this config.
`rootModule.configure(config);`
https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/Main.java#L550

Each thread can create config completely isolated modules tree from each other and use them.
We are going to have pool of thread, so tree of modules can be shared between execution on file. New file is passed to thread to process, we just process new file, tree-of-modules  stay the same.

@rnveach , ping.

**Comment by rnveach** [comment]
It is because different threads will need different instances in each thread. Each thread may get the same list of modules, but the clone service controls which instances are unique and which are duplicated in the threads.

Modules annotated with `GlobalStatefulCheck` need to give every thread the same instance as they are multi-file and must examine every single file. Each unique thread is a different unique file. `StatelessCheck` is very similar except it doesn't declare the module to be multi-file and just says it is safe for every thread to use the same instance. I don't remember if there was another, unique reason for this annotation, but it just means the module is very simple. `FileStatefulCheck` just means that each thread gets it's own unique, instance. This last one is how every unmarked check will be handled.

> Each thread can create config completely isolated modules tree from each other and use them

Not for multi-file modules. They must examine every file and each thread will be working on a different file.

**Comment by romani** [comment]
>  but the clone service controls which instances are unique and which are duplicated in the threads.

lets first of all become to understand why "clone" is required ???
We have config file, we just need to create tree-of-modules, each module can have its own instance for each thread. We have 150 modules, is we use 4 threads, we try to economize creation of 600 objects? 
Did somebody do time/memory estimation to prove reason of such optimization ?
 
> GlobalStatefulCheck

Still see them as hack that should not support for now. Annotation can stay, but it should move execution of them out of all. 

> StatelessCheck

I still not sure on how we will use this marker. We will use it only on Treewalker lever when we will try to run all StatelessCheck in one thread withtout waiting for response on visitToken. We will think on how to make it later on. As for now we should treat them as FileStatefulCheck and run the same. 

> FileStatefulCheck 

This is what we will use in MTreeWalker. As AST parsing is done, we can split all Checks in several groups for traversing in parallel (for each sub-thread) and run all of them. Wait for all of them completion, apply filters, return result.

**Comment by rnveach** [comment]
These annotations are not dependent on TreeWalker or not. They mean the same thing to FileSets and AbstractChecks.

> I still not sure on how we will use this marker

Marker is already used. The only checks not marked are GlobalStateful as far as I know.

> Treewalker level when we will try to run all StatelessCheck in one thread withtout waiting for response on visitToken

1 thread cannot execute multiple methods parallel. Either it is single threaded completely or it is multi-threaded. There is no reason these states need to be run in a single threaded mode. They are guaranteed not to be mutli-file and majority doesn't store a temporary state in a field (see `ArrayTypeStyleCheck`). They are the simplest of the simply checks.
I assume it just exists to separate it from the horror that is `GlobalStateful`.

**Comment by romani** [comment]
> Marker is already used.

they are in code, to mark all Checks what they are, that is good marker in all cases.
But MT implementation now in big question. So usage of them might be different.

> 1 thread cannot execute multiple methods parallel.

Ones we come to this, we have choice to run all StatelessCheck together with FileStatefulCheck OR look at lightweight threads(fibers) implementations and run each visitToken on its own Instance of Checks.
When we come to this, we will experiment. Creation of standard thread for each Check is too heavy, it is is better to run all of them in on thread.
NO need to economize memory on instances of StatelessCheck, 1 instance vs 4 instances is not a big deal.

**Comment by romani** [comment]
Looks like multi-file support is problematic case in MT design.

Each FileSetCheck is NOT able to do multiple file processing: `void beginProcessing();` `SortedSet<LocalizedMessage> process(...)` `void finishProcessing();` , validation result is on only at `process` method (that is executed by on each file)
https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/api/FileSetCheck.java#L49

TranslationCheck do huge hack for [reporting violations](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/TranslationCheck.java#L325)  

[getMessageDispatcher()](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java#L104) method is used only in TranslationCheck. This method is like this already 16 years. The same protected visibility is for `protected final void fireErrors(String fileName)` method in AbstractFileSetCheck, and just for TranslationCheck.
We invested so much time in Translation validation, it is really useful Check.
BUT we need to keep this HACK out of main stream for now. 

JavadocPackageCheck will also not be suitable in model when different Check instances are used in each thread, or additional mess with making his field static to be shared between instances. It is targeting uncovered by Checkstyle area - folder validation. It is better to process it too as second execution. I even ok to remove this check at all, or move it to sevntu. OR keep global annotation on it, and keep make Checkstyle smarter a bit to keep same instance in all threads.

--------------------

As an option we can leave them to be compatible only with Checker parent module not a MultiThreadChecker, so use who need them will have to run them in separate execution/config that even worse then embedded second execution.

This two Checks are probably good example of what happening in thirparty Checks, I do believe there even more ugly cases could be found that actually do their work.

---------------

so delema is:
- in model "in each thread separate instance" - all Checks stay threading context unaware, annotations are required to split global and file-scoped, if no annotation we can think that all FileSetsChecks are global and AbstractChecks are FileStateful. Multi-file support only with knowledge of Checkstyle threading model and Checkstyle keep same instance for each thread.

- in model "shared instances of Checks between threads" - all Checks should kind of be aware of threading 
expect, annotations are mandatory. Multi-file support only with knowledge of Checkstyle threading model.

**Comment by romani** [comment]
> Each thread was a single file, and thanks to a cloning service, multi-file checks were the same instance across all threads, since they need to examine all files.

I still do not understand reason of cloning service. Checker should create tree of modules(separate instances) as much time as it is required, manage `GlobalStateful` to be the same in all trees. As firs update to make Checker as MT, as second update Treewalker as MT , all other optimizations and extra MT stuff could be done later. Checkstyle need to have MT config by default to always run over it and make as much testing as possible. CI should continue to have single-threaded Checker confirg to make sure that all works fine. As this become stable implementation, we can  start think of merging of MTChecker to Checker or ..... .

**Comment by rnveach** [comment]
Even if checker will control the instances, you still need some type of cloning service somewhere to control modules instannces. Multi-file modules must be a single instance throughout all threads, as they must interact with every file and build a repository of data. Non-multi-file modules must be different instances in each thread. You will need something to say how instances are controlled for each module.
It is also better that such a new util be outside Checker as it will make Checker too complex to contain all this.

**Comment by romani** [comment]
> Even if checker will control the instances, you still need some type of cloning service somewhere to control modules instances. 

Cloning is not required, I try to avoid this word. We have simple rules: if Stateless or FileStateful then create by config (as usually). If Globalstateful then reuse instance that already exists. So "create or reuse" but not a clone.

> It is also better that such a new util be outside Checker as it will make Checker too complex to contain all this.

I have no problems to not host such logic in Checker, some other class can manage creation of instances and keep cache of GlobalStateful instances.

**Comment by rnveach** [comment]
This is what the PR already does then.

**Comment by romani** [comment]
@rnveach ,

I do want to completely avoid term "clone" , PR need to be redone.
This PR is not in merge-able state.

If we are on the same page on how to make it, lets close this PR and update issue(probably creation of new issue, without history and discussions) with details on what is required to let smb open new PR with implementation we are ok.

**Comment by rnveach** [comment]
@romani Besides the name, what else makes you say the PR needs to be redone?

**Comment by romani** [comment]
I am not sure about new interfaces like `OneCheckInstancePerThread`.
changes in API is also suspicious , we should avoid this, but I will review it one more time.
I think coverage will be required, ...

**Comment by romani** [comment]
#4882 is closed, but should be reviewed as it contains valuable discussion.
One day we will fins time to come back to MT.
